packages/vite/src/node/build.ts

```diff
@@ -875,13 +889,66 @@ async function buildEnvironment(
       prepareOutDir(resolvedOutDirs, emptyOutDir, environment)
     }
 
-    const res: RolldownOutput[] = []
-    for (const output of normalizedOutputs) {
-      res.push(await bundle[options.write ? 'write' : 'generate'](output))
-    }
+    const res = await build()
+
     logger.info(
       `${colors.green(`✓ built in ${displayTime(Date.now() - startTime)}`)}`,
     )
+
+    async function build() {
+      const res: RolldownOutput[] = []
+      for (const output of normalizedOutputs) {
+        // TODO(underfin): using the generate at development build could be improve performance.
+        res.push(await bundle![options.write ? 'write' : 'generate'](output))
+      }
+
+      if (server) {
+        // watching the files
+        for (const file of bundle!.watchFiles) {
+          if (path.isAbsolute(file) && fs.existsSync(file)) {
+            server.watcher.add(file)
+          }
+        }
+
+        // Write the output files to memory
+        for (const output of res) {
+          for (const outputFile of output.output) {
+            server.memoryFiles[outputFile.fileName] =
+              outputFile.type === 'chunk' ? outputFile.code : outputFile.source
+          }
+        }
+      }
+      return res
+    }
+
+    if (server) {
+      server.watcher.on('change', async (file) => {
+        const startTime = Date.now()
+        const hmrOutput = (await bundle!.generateHmrPatch([file]))!
+        if (hmrOutput.patch) {
+          const url = `${Date.now()}.js`
+          server.memoryFiles[url] = hmrOutput.patch
+          server.ws.send({
+            type: 'update',
+            updates: hmrOutput.hmrBoundaries.map((boundary) => {
+              return {
+                type: 'js-update',
+                url,
+                path: boundary.boundary,
+                acceptedPath: boundary.acceptedVia,
+                timestamp: 0,
+              }
+            }),
+          })
+          logger.info(
+            `${colors.green(`✓ Found ${path.relative(root, file)} changed, rebuilt in ${displayTime(Date.now() - startTime)}`)}`,
+          )
+
+          await build()
```

Comment:
Maybe this `build` should be debounced?

Reply:
Yeah. Follow https://github.com/vitejs/rolldown-vite/pull/107#discussion_r2053232740, it should be debounced to batch multiple files changes.

Reply:
Added debounced.